### DISCUSSION TITLE
Could not initialize class DataWriterTAOPeer

### DISCUSSION THREAD

**Discussion by axuchao458** [OpenDDS/OpenDDS#4872] [Q&A]
hello everyone:
I am a Java developer currently using OpenDDS to communicate with a simulation software and have encountered some issues.

1. If I want to communicate with the simulation software, I need to use the OpenDDS compiled jar and DLL provided by the simulation software. If I use other versions of OpenDDS jar and DLL compiled by myself, I cannot connect to DDS. I consulted them, and their response was that the version must be consistent, otherwise the connection cannot be made. They provided a 3.17-dev version instead of a release version. Is this necessary? Is it necessary to have a consistent version for communication?
2. After I connect to DDS, there is always a situation where messages are lost. That is, after I send them, the simulation software cannot receive them, but there is no error in the logic I send. I don't know how to troubleshoot this problem?
3. My code will encounter an error such as "Could not initialize class YCMod. _YCDestoryElementTruthEventDataWriterTAOPeer" because this error is a native method, and I am not quite sure what is going on?
![image](https://github.com/user-attachments/assets/d9ed30b3-638d-4a9b-a74e-a9002f237c5b)
I always receive a prompt that the DLL has already been loaded by a thread. In order to handle this error, I used the same thread to execute my publish and subscribe logic, but still encountered this error. I don't know how to handle it
![image](https://github.com/user-attachments/assets/ec91d31c-abe5-4bde-9221-d45e31a1c79a)
This is the logic in my code, I retrieve the dataWriter
![image](https://github.com/user-attachments/assets/07f24e1e-3614-45ba-ab58-f1973088106f)
This is the specific implementation of the method, and my breakpoint will go to the sentence in the red box
![image](https://github.com/user-attachments/assets/31e4028f-80b1-47a8-947f-9db723ef0fc6)
I will eventually reach a native method, and then I won't be able to break points

Thank you all for your help

**Comment by iguessthislldo** [comment]
> They provided a 3.17-dev version instead of a release version. Is this necessary? Is it necessary to have a consistent version for communication?

It shouldn't in most cases, but 3.17 was released 3 years ago, so it might have some issues depending on your setup, particularly if XTypes is being used. If I can't really say much more without more details like a debug log (can be made using command line arguments like `-DCPSDebugLevel 10 -ORBLogFile debug.log`) or a [Wireshark capture](https://objectcomputing.com/resources/publications/sett/october-2021-wireshark-and-opendds).

> After I connect to DDS, there is always a situation where messages are lost. That is, after I send them, the simulation software cannot receive them, but there is no error in the logic I send. I don't know how to troubleshoot this problem?

Is it not receiving all messages you send or just some of them? Again, if you attach a debug log from both sides would be helpful.

> My code will encounter an error such as "Could not initialize class YCMod. _YCDestoryElementTruthEventDataWriterTAOPeer" because this error is a native method, and I am not quite sure what is going on?

I'm not familiar with this particular aspect of Java, but it sounds like [another part of the process is trying to use the same native library](https://stackoverflow.com/questions/36936948/java-lang-unsatisfiedlinkerror-native-library-xxx-so-already-loaded-in-another). Is this simulation software you're using a separate process or is it a library in the same process? If it's the same process then that might explain why this error is happening. It would also be why they want you to use the same 3.17 version as them because I'm not sure we support two different OpenDDS versions coexisting in the same program.

**Reply by axuchao458** [reply]
Thank you for your assistance. The project team I'm currently collaborating with isn't very proactive, making it challenging to coordinate joint debugging with them. Nevertheless, they've provided me with a new configuration file. Since using it, there has been significant improvement in the loss situation. Although occasional losses still occur, I'll share the configuration file here.
[dds.txt](https://github.com/user-attachments/files/18471587/dds.txt)
